File: android_os/app/src/main/java/com/mrkhntr/workscreentime/notify/WebhookSender.kt
package com.mrkhntr.workscreentime.notify

import okhttp3.Call
import okhttp3.Callback
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import okhttp3.Response
import org.json.JSONObject
import java.io.IOException

/// Sends an already-rendered webhook request from the core's `sendWebhook`
/// effect. The core builds url/headers/body; native only performs the POST.
class WebhookSender {
    private val client = OkHttpClient()
    private val jsonType = "application/json".toMediaType()

    fun send(request: JSONObject) {
        val url = request.optString("url")
        if (!url.startsWith("http://") && !url.startsWith("https://")) return

        val body = request.optJSONObject("body") ?: JSONObject()
        val builder = Request.Builder().url(url).post(body.toString().toRequestBody(jsonType))

        request.optJSONObject("headers")?.let { headers ->
            val keys = headers.keys()
            while (keys.hasNext()) {
                val key = keys.next()
                builder.header(key, headers.optString(key))
            }
        }

        runCatching {
            client.newCall(builder.build()).enqueue(object : Callback {
                override fun onFailure(call: Call, e: IOException) {}
                override fun onResponse(call: Call, response: Response) {
                    response.close()
                }
            })
        }
    }
}

File: mac_os/Sources/WorkScreenTimeApp/WebhookSender.swift
import Foundation

/// Sends an already-rendered webhook request from the core's `sendWebhook`
/// effect. The core builds the URL/headers/body; native only performs the POST.
struct WebhookSender {
    func send(_ request: [String: Any]) {
        guard
            let urlString = request["url"] as? String,
            let url = URL(string: urlString),
            let scheme = url.scheme?.lowercased(),
            ["http", "https"].contains(scheme)
        else {
            return
        }

        var urlRequest = URLRequest(url: url)
        urlRequest.httpMethod = (request["method"] as? String) ?? "POST"

        if let headers = request["headers"] as? [String: Any] {
            for (name, value) in headers {
                urlRequest.setValue("\(value)", forHTTPHeaderField: name)
            }
        }

        if let body = request["body"] as? [String: Any],
           JSONSerialization.isValidJSONObject(body),
           let data = try? JSONSerialization.data(withJSONObject: body) {
            urlRequest.httpBody = data
        }

        URLSession.shared.dataTask(with: urlRequest).resume()
    }
}
